src/v/v8_engine/script_dispatcher.h

```diff
@@ -0,0 +1,105 @@
+ */
+
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "model/metadata.h"
+#include "utils/mutex.h"
+#include "v8_engine/script.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/util/variant_utils.hh>
+
+#include <absl/container/node_hash_map.h>
+#include <absl/hash/hash.h>
+
+#include <exception>
+#include <iostream>
+#include <variant>
+
+namespace v8_engine {
+
+struct script_compile_params {
+    ss::sstring function_name;
+    iobuf code;
+    size_t max_heap_size_bytes{10000};
+    size_t run_timeout_ms{200};
+};
+
+// This class implement database for v8_runtimes. It can be updated by
+// data_policy_manager. Fetch request will only run get for getting v8_runtimes
+template<typename Executor>
+class script_dispatcher {
+    constexpr static size_t _mutex_count{50};
+
+public:
+    explicit script_dispatcher(Executor& executor)
+      : _mutexes(_mutex_count)
+      , _executor(executor) {}
+
+    ss::future<>
+    update(model::topic_namespace& topic, script_compile_params script_params) {
+        auto hash = absl::Hash<model::topic_namespace>{}(topic);
+        auto lock = co_await _mutexes[hash % _mutexes.size()].get_units();
+        _scripts.insert_or_assign(topic, std::move(script_params));
+    }
+
+    ss::future<> remove(model::topic_namespace topic) {
+        auto hash = absl::Hash<model::topic_namespace>{}(topic);
+        auto lock = co_await _mutexes[hash % _mutexes.size()].get_units();
+        _scripts.erase(topic);
+    }
+
+    /// This method return v8_runtimes for data-policy (aka topic on curent
+    /// moment). If hash_table contains compile params get needs to compile this
+    /// script, so on this moment we need to lock hash_table. Also if sbmd try
+    /// to get v8_runtimes it needs to wait compilation and do not compile twice
+    ss::future<ss::lw_shared_ptr<script>> get(model::topic_namespace topic) {
+        auto hash = absl::Hash<model::topic_namespace>{}(topic);
+        auto lock = co_await _mutexes[hash % _mutexes.size()].get_units();
+
+        auto it = _scripts.find(topic);
+        if (it == _scripts.end()) {
+            co_return nullptr;
+        }
+
+        co_return co_await ss::visit(
+          it->second,
+          [this, topic](script_compile_params& params) mutable
+          -> ss::future<ss::lw_shared_ptr<script>> {
+              ss::lw_shared_ptr<script> new_runtime
+                = ss::make_lw_shared<script>(
+                  params.max_heap_size_bytes, params.run_timeout_ms);
+
+              co_await new_runtime->init(
+                std::move(params.function_name),
+                std::move(params.code),
+                _executor);
+
+              _scripts.insert_or_assign(topic, new_runtime);
+              co_return new_runtime;
+          },
+          [](ss::lw_shared_ptr<script>& script_ptr)
+            -> ss::future<ss::lw_shared_ptr<script>> { co_return script_ptr; });
+    }
+
+private:
+    using value
```

Comment:
I don't think this variant would be needed. If the points above are taken, if the caller waits on a value in the `_mutexs` collection, by the time it is finished, the script is finished compiling and it will observe the script.